### PR TITLE
CLDR-15954 Add test of unit preferences test

### PR DESCRIPTION
**Pull request by macchiati**
CLDR-15954

The main goal was to add tests of test files, for verification. It grew a bit as I discovered and fixed problems.

common/testData/units/unitLocalePreferencesTest.txt
- new test file to check problems in following the spec with locales

common/testData/units/unitPreferencesTest.txt
common/testData/units/unitsTest.txt
- New SPDX id
- Fixed the pointer to generation program.

tools/cldr-code/src/main/java/org/unicode/cldr/tool/GenerateUnitTestData.java
- addition of 'public void generateUnitLocalePreferences() {' to generate the new test data file
- and some utility methods
- refactoring a bit

tools/cldr-code/src/main/java/org/unicode/cldr/util/GrammarInfo.java
- keep the current behavior for which units get grammar.
- needed because of a fix in XXX

tools/cldr-code/src/main/java/org/unicode/cldr/util/Rational.java
- some general cleanup of Rational, for issues noticed in building test data
- Also made Rational inherit from Number.
- NaN was not behaving as expected. Reorganized so that the constructor is simple, and the Rational.of() does more checks
- That also touches add(), subtract(), ... which are made simpler and more robust
- Changed toBigDecimal(), doubleValue() to be cleaner; added floatValue()
- Added utilities of(double), intValue(), longValue(), approximatelyEquals(Number)

tools/cldr-code/src/main/java/org/unicode/cldr/util/UnitConverter.java
- the code for getSystemsEnum(String unit) wasn't following the comments, thus getting the wrong answer for some compound units.
- added a format method, factoring out code that needed to be in various places.

tools/cldr-code/src/main/java/org/unicode/cldr/util/UnitPreferences.java
- made the fast map be a safe singleton (well, memoized):
    - Supplier<Map<String, Map<String, Multimap<String, UnitPreference>>>> quantityToUsageToRegionToInfo = Suppliers.memoize(() -> getRawFastMap());
- fixed the smallest value to be zero, so that the algorithm to get the preferences didn't have to special-case the least item.
- moved code from test/generation to here for getting unitPreferences
- added API to get all the quantities

tools/cldr-code/src/main/java/org/unicode/cldr/util/Units.java
- fixed get short/long to work for either case, because MeasureUnit may or might not have a prefix.

tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestUnits.java
- Added Rational tests for NaN, INF
- Added testQuantitiesMissingFromPreferences(), to identify when quantities we don't have in preferences
- Added test of the unitPreferencesTest.txt, unitsTest.txt, and (new) unitLocalePreferencesTest.txt data files
- Also added a testUnitLocalePreferencesTestIcu() so that we could see how ICU was doing without waiting for integration (it is under a flag, so not normally run).

- [ ] This PR completes the ticket.

ALLOW_MANY_COMMITS=true

### CODE REVIEW

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/GrammarInfo.java
Comment (srl295) [inline tools/cldr-code/src/main/java/org/unicode/cldr/util/GrammarInfo.java:759]: need a TODO ticket link for this?

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/UnitConverter.java
Comment (srl295) [inline tools/cldr-code/src/main/java/org/unicode/cldr/util/UnitConverter.java:2266]: better not to propagate locale string slicing

```suggestion
    final String localeBase = ULocale.forLanguageTag(languageTag).getBaseName();
```

something like that?

**Review by srl295 — COMMENTED** [review]
still reviewing

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/GrammarInfo.java
Comment (macchiati) [reply tools/cldr-code/src/main/java/org/unicode/cldr/util/GrammarInfo.java:759]: Filed yesterday: https://unicode-org.atlassian.net/browse/CLDR-17459

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/UnitConverter.java
Comment (macchiati) [reply tools/cldr-code/src/main/java/org/unicode/cldr/util/UnitConverter.java:2266]: I'm planning on another PR anyway, and will look at this. Since the beta is already tagged, would you mind approving now?

File: common/testData/units/unitPreferencesTest.txt
Comment (btangmu) [inline common/testData/units/unitPreferencesTest.txt:2]: Add a comment, "DO NOT EDIT THIS FILE, it is autogenerated by GenerateUnitTestData.java"?

File: common/testData/units/unitsTest.txt
Comment (btangmu) [inline common/testData/units/unitsTest.txt:1]: Add a comment, "DO NOT EDIT THIS FILE, it is autogenerated by GenerateUnitTestData.java"?

File: tools/cldr-code/src/main/java/org/unicode/cldr/tool/GenerateUnitTestData.java
Comment (btangmu) [inline tools/cldr-code/src/main/java/org/unicode/cldr/tool/GenerateUnitTestData.java:204]: TODO comment should have a ticket link, ticket shouldn't be closed until comment is removed

File: tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestUnits.java
Comment (btangmu) [inline tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestUnits.java:4385]: TODO should have ticket number

File: tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestUnits.java
Comment (btangmu) [inline tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestUnits.java:4435]: TODO should have ticket and closing parenthesis

**Review by btangmu — APPROVED** [review]
rubber-stamp approval; added suggestions for improvement

File: tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestUnits.java
Comment (macchiati) [reply tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestUnits.java:4385]: I'd like to do that in a subsequent PR. I had to squash, and it is much simpler to do in a follow-on PR.

**Comment by macchiati** [comment]
I will, however, be waiting until I get the go-ahead from Dragan

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/GrammarInfo.java
Comment (srl295) [reply tools/cldr-code/src/main/java/org/unicode/cldr/util/GrammarInfo.java:759]: good, should ideally be in code

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/UnitPreferences.java
Comment (srl295) [inline tools/cldr-code/src/main/java/org/unicode/cldr/util/UnitPreferences.java:339]: should be data driven.. or actually lookup prefernces

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/UnitPreferences.java
Comment (srl295) [inline tools/cldr-code/src/main/java/org/unicode/cldr/util/UnitPreferences.java:325]: function should move into supplemental data info

**Review by srl295 — APPROVED** [review]
OK for now, comments noted

**Pull request merged**
